htmake/htmake.1
.TH HTMAKE 1 "27 March 1999" "Debian GNU/Linux" "GLOBAL cgi Utilities"
.\" Copyright (c) 1999, 2000  Ron Lee.  All rights reserved.
.\" This text may be distributed under the same terms as the htmake source.
.\" All copies of this manual must acknowledge the author and include the
.\" above copyright statement.
.\"
.\" In cases of discrepancy the source is authorative.  Duh 8^}
.SH NAME
htmake \- create GLOBAL hypertext source searchable by gsearch.cgi
.SH SYNOPSIS
.BR htmake " [" "htags options" "] [" --url=base-url ]
.RI [ destdir ]
.SH "DESCRIPTION"
.B htmake
is a wrapper script that creates searchable hypertext from program 
source code and records information into a \fBpath database\fP
needed for a centralised search mechanism.
.PP
Prior to using this command you must have installed gsearch.cgi
using the \fBhtconfig -I\fP option and configured the location
of the \fBpath database\fP in \fIhtmake.conf\fP
.PP
Before you can create the hypertext source you need to execute
\fBgtags\fP(1) from the root directory of the source tree.
\fBhtmake\fP is then called from the same place.
.PP
\fBhtmake\fP uses \fBhtags\fP(1) to create the hypertext with a
unique id and \fBhtconfig\fP(1) to make the required entries in
the \fBpath database\fP.
.SH OPTIONS
The following options will be passed to \fBhtags\fP(1) if used.
.TP
.B -a
Make an alphabetical function index.
.TP
.B -c
Compress html. Use gzip(1) with .ghtml suffix.
.TP
.B -C
Compress html. Use gzip(1) with .html.gz suffix.
.TP
.B -d \fItagdir\fP, --tags=gtagsdbpath
Specifies the directory in which GTAGS and GRTAGS exist. The default is the current directory.
.TP
.B -F
Use frames for definition index, file index and main view.
.TP
.B -l
Make a name tag for each line of hypertext created.
.TP
.B -n
Print out the line numbers.
.TP
.B -t title
The title of this hypertext. Defaults to the last component of the current path.
.TP
.B --action=cgi-url
Specify the location search queries are submitted to.
Defaults to the value set in \fBhtmake.conf\fP or to
.br
.I http://localhost/cgi-bin/gsearch.cgi
.PP
The remaining options are more general or are passed to \fBhtconfig\fP(1).
.TP
.B --url=base-url
the base url for your new hypertext source.
For best results always use absolute url's here.
.br
eg. \fIhttp://host/~user/mysource\fP
.br
It defaults to \fIfile:/destdir\fP
.TP
.I destdir
The directory in which hypertext is generated.
The default is \fI./HTML\fP
.TP
.B -v, --verbose
Activate the inbuilt screen saver. Give your phosphors a workout.
.SH EXAMPLES
$ cd /usr/src/myprog
.br
$ gtags
.br
$ htmake
.PP
$ cd /usr/src/myprog
.br
$ gtags -c
.br
$ htmake \-c \-\-url=http://host/~user/prog/HTML \\
  /home/user/public_html/prog
.SH FILES
.I /etc/gtags/htmake.conf
.br
.I path database
.SH BUGS
As a wrapper \fBhtmake\fP gets to inherit bugs from other software too.
Isn't life fun.
.PP
This is \fIalpha\fP software \- expect anything!
.SH "SEE ALSO"
.BR htconfig (1), " htags" (1), " gtags" (1), " global" (1).
.SH AUTHOR
Ron Lee <ron@debian.org>
.SH CREDITS
Thanks to Shigio Yamaguchi <shigio@gnu.org> for creating
\fBGLOBAL\fP and releasing it as free software.
.br
A truly cool productivity tool!


htmake/htmake.pl
#!/usr/bin/perl
#
# Copyright (c) 1999, 2000
#		Ron Lee <ron@debian.org> All rights reserved.
#
#  You may distribute this software under the terms of the GNU GPL version 2
#  or later.
#
#	htmake.pl					25-Mar-99
#
#	htmake is a wrapper that calls htags(1) with an --id attribute
#	and runs htconfig on the generated hypertext.
#
# OPERATION:	generate unique id.
#		call htags to make html source
#		call htconfig to update database.
#		must be called from src dir.
#
#	htmake [htags options] --action=cgi-action --url=base-url --tags=gtagsdbpath [destdir]
#
#	TODO: add --rebuild option, to recycle old id. ??
#

$com = $0;
$com =~ s/.*\///;
$usage = "usage: $com [-a][-c|-C][-h][-l][-n][-v][-d tagdir, --tags=tagdir][-t title]\n";
$usage .= "         [--action=cgi-url][--url=base-url][destdir]\n";

#-----------------------------------------------------------------
# DEFAULTS
#-----------------------------------------------------------------

$CONFIG = '/etc/gtags/htmake.conf';
$action = 'http://localhost/cgi-bin/gsearch.cgi';

#----------------------------------------------------------------
# UTILITIES
#----------------------------------------------------------------

sub getcwd {
	local($dir) = `/bin/pwd`;
	chop($dir);
	$dir;
}
sub realpath {
	local($dir) = @_;
	local($cwd) = &getcwd;
	chdir($dir) || &error("directory '$dir' not found.");
	$dir = &getcwd;
	chdir($cwd) || &error("Cannot return to '$cwd'.");
	$dir;
}
sub usable {
	local($command) = @_;
	foreach (split(/:/, $ENV{'PATH'})) {
		return 1 if (-x "$_/$command");
	}
	return 0;
}
sub error {
	print STDERR "$com: $_[0]\n\n";
	exit 1;
}
sub usage {
	print STDERR $usage;
	exit 1;
}

#-----------------------------------------------------------------
# MAIN
#-----------------------------------------------------------------

&usable('htags') or &error("htags function not availiable.");
&usable('htconfig') or &error("htconfig function not availiable.");
-f $CONFIG and do $CONFIG or &error("Cannot read $CONFIG, see htconfig --install option first.");
&error("GSPATHDATA must be defined in $CONFIG") unless $GSPATHDATA;
$action = $ACTION if $ACTION;
$destdir = $url = $tags = '';
$htagsflags = 'f';
%longoption = (	action => '/cgi-bin/gsearch.cgi', url => '', tags => '', verbose => 1, );
while ($ARGV[0] =~ /^-/) {
	$opt = shift;
	if ($opt =~ /^--(\w+)=?([-+.:~\/\w]*)/) {
		&usage unless exists $longoption{$1};
		$$1 = $2 || $longoption{$1};
		next;
	}
	&usage if ($opt =~ /[^-acCdFlntv]/);
	if ($opt =~ /a/) { $htagsflags .= 'a'; }
	if ($opt =~ /c/) { $htagsflags .= 'c'; }
	if ($opt =~ /C/) { $htagsflags .= 'C'; }
	if ($opt =~ /F/) { $htagsflags .= 'F'; }
	if ($opt =~ /l/) { $htagsflags .= 'l'; }
	if ($opt =~ /n/) { $htagsflags .= 'n'; }
	if ($opt =~ /v/) { $verbose = 1; }
	if ($opt =~ /d/) {
		$opt = shift;
		last if ($opt eq '');
		$tags = $opt;
	} elsif ($opt =~ /t/) {
		$opt = shift;
		last if ($opt eq '');
		$title = $opt;
	}
}
$destdir = &realpath($ARGV[0]) if $ARGV[0];

# get unique id
#
&getcwd =~ /([^\/]*)$/ and $id = $1;
$id =~ s/[^\w]//g;
$id .= time;

# call htags
#
$htagsflags .= 'v' if $verbose;
@htagsopt = ("-$htagsflags");
push @htagsopt,('-t',"$title") if $title;
push @htagsopt,('-d',"$tags") if $tags;
print "Executing: htags @htagsopt --action=$action --id=$id --nocgi $destdir\n" if $verbose;

# htags is a C program now, so the number of arguments is important.
#
if ($destdir eq '') {
	system('htags',@htagsopt,"--action=$action","--id=$id","--nocgi") == 0 or
		&error("Failed to execute htags.");
} else {
	system('htags',@htagsopt,"--action=$action","--id=$id","--nocgi","$destdir") == 0 or
		&error("Failed to execute htags.");
}

# call htconfig
#
@htconfigopt =();
push @htconfigopt,('-v') if $verbose;
push @htconfigopt,("--url=$url") if $url;
push @htconfigopt,("--tags=$tags") if $tags;
print "Executing: htconfig @htconfigopt $destdir\n" if $verbose;
system('htconfig',@htconfigopt,"$destdir") == 0 or &error("Failed to execute htconfig.");

exit 0;
